Move imagejob deletion to imagelist controller

pmengelbert opened the pull request:
**This PR is based on #204, please see the relevant diff here:**
https://github.com/pmengelbert/eraser/compare/fix_potential_reconciliation_bug...pmengelbert:eraser:move_imagejob_deletion_to_imagelist_controller

**What this PR does / why we need it**:
- Delete imagejob in imagelist controller, not from the imagejob controller.
- Update Imagelist status field in the imagelist controller, not in the imagejob controller.

This is necessary because:
- The `imagejob` is being repurposed for management of `collector` pods, and not just `eraser` pods.
- The `imagecollector controller` needs to monitor the `imagejob` to determine when it is completed, in order to proceed to the scanning phase.
- With the current behavior of the `imagejob controller`, the `imagejob` will be cleaned up once the `eraser` pods are finished, preventing the `imagecollector controller` from properly monitoring the `imagejob`.
- This issue will block further progress on the `imagecollector controller`, so it should be prioritized.

**Which issue(s) this PR fixes** *(optional, using `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when the PR gets merged)*:
Fixes #200 
Based on top of PR #204, that one should be merged first

**Special notes for your reviewer**:
Please enjoy, and see commit messages for rationale and details

pmengelbert committed:
Check for child ImageJobs in ImageList Controller

- Add a new watch to the controller. This will be updated to correctly
  filter events; it is currently a placeholder to prevent perpetual
  loops.
- In the reconciliation body, when creating an ImageJob, label it so we
  can later determine whether it belongs to the imagelist controller or
  the imagecollector controller.
- In the reconciliation body, check for imagejobs that belong to the
  imagelist controller. If none, proceed to creation. If one child
  imagejob, then handle it (currently a stub). If multiple child
  imagejobs, then we have encountered an error state.
- Create `consts` package to hold constants that the various controllers
  will share

Signed-off-by: Peter Engelbert <pmengelbert@gmail.com>

pmengelbert committed:
Move deletion of ImageJobs to ImageJob Controller

- Create new `controllers/util` package for functions that are shared
  between controllers
- Add predicate functions to both ImageList Controller and ImageJob
  controller. These predicates ensure that if an ImageJob's Status.Phase
  is Completed or Failed, the update event will be handled by the
  controller that created it, and not the ImageJob controller.
- Handle deletion in the ImageList controller for ImageJobs created by
  that controller
- I've run the e2e tests and everything was fine there
- I've manually tested for delayed deletion when the ImageJob enters the
  Completed state, but I didn't have an easy way to test when the
  ImageJob enters a Failed state.

Signed-off-by: Peter Engelbert <pmengelbert@gmail.com>

cpuguy83 commented on controllers/imagelist/imagelist_controller.go:
```diff
@@ -86,19 +92,55 @@ func (r *Reconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Resu
 			OwnerReferences: []metav1.OwnerReference{
 				*metav1.NewControllerRef(imageList, imageList.GroupVersionKind()),
 			},
+			Labels: map[string]string{consts.ImageJobOwnerLabelKey: imageList.Name},
```
I think OwnerReference does what we need here?

pmengelbert replied:
Is there a way to select for that in the call to `r.List(...)` when we are listing the existing `ImageJob`s? I couldn't figure that out so I went with the labels instead. I am looking at the `sigs.k8s.io/controller-runtime/pkg/client.ListOptions` struct, and I don't see any option to select by `OwnerReference`.

The other option is to `List` all `ImageJob`s and filter them manually in the controller, is that the better way to go?

pmengelbert replied:
Filtering client-side using `UID`s looks like the way to go, and more reliable than just using the owner's `Name`. Will mark resolved once tests pass

pmengelbert commented on controllers/imagelist/imagelist_controller.go:
```diff
@@ -86,19 +92,54 @@ func (r *Reconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Resu
+	case 0:
+		return r.handleImageListEvent(ctx, &req, &imageList)
+	case 1:
+		return r.handleJobListEvent(ctx, &imageList, &jobList.Items[0])
```
This needs additional logic. In addition to an ImageJob being moved to the `Completed` state, we will **also** reach this point when both of the following conditions are true:
1. Reconcile was triggered by an `UpdateEvent` to the `ImageList`
2. There is already an `ImageJob` in progress, and it is not yet complete.

This is related to #203 , and the best thing to do in that case is probably to requeue the `Imagelist`'s `UpdateEvent` for some reasonable time in the future, and to keep punting it until there are no `ImageJob`s in the running state.

pmengelbert committed:
Add logic for updates to imagelist during imagejob

In addition to an ImageJob being moved to the Completed state, we will
**also** reach this point when both of the following conditions are true:

1. Reconcile was triggered by an UpdateEvent to the ImageList
2. There is already an ImageJob in progress, and it is not yet complete.

This is related to #203 , and the best thing to do in that case is
probably to requeue the Imagelist's UpdateEvent for some reasonable time
in the future, and to keep punting it until there are no ImageJobs in
the running state.

Signed-off-by: Peter Engelbert <pmengelbert@gmail.com>

pmengelbert committed:
Use OwnerReference.UID to filter out ImageJobs not owned by ImageList

Signed-off-by: Peter Engelbert <pmengelbert@gmail.com>

pmengelbert committed:
Pass a pointer to avoid large copy

Signed-off-by: Peter Engelbert <pmengelbert@gmail.com>

pmengelbert committed:
Fix typo

Signed-off-by: Peter Engelbert <pmengelbert@gmail.com>

pmengelbert committed:
Remove label on child imagejobs

Signed-off-by: Peter Engelbert <pmengelbert@gmail.com>

ashnamehrotra commented:
Can we also move the functionality from lines 287-310 into the imagelist controller? When we use the imagejob to schedule collector pods we won't need a reference to any imagelist

pmengelbert commented:
@ashnamehrotra Yes, that's a very good idea. I was running into problems with this yesterday and the conditional logic was getting messy. I'll work on that now

pmengelbert committed:
Decouple ImageJob from ImageList

Signed-off-by: Peter Engelbert <pmengelbert@gmail.com>

pmengelbert commented:
@ashnamehrotra done

pmengelbert committed:
Remove status update

This must have been the byproduct of a rebase somewhere.

Signed-off-by: Peter Engelbert <pmengelbert@gmail.com>

sozercan reviewed (approved):
LGTM

cpuguy83 reviewed (approved):
LGTM
